lib/tools.py

```diff
@@ -160,3 +152,235 @@ def reset_retriever(scope):
                 os.remove(os.path.join(HOME_DIR, 'connections.config'))
             except:
                 pass
+
+
+def is_valid_json(input_file):
+    """Validate a json file"""
+    try:
+        fp = open(input_file, 'r')
+        json.loads(fp.read())
+    except ValueError:
+        return False
+    return True
+
+
+def json2csv(input_file, output_file=None, header_values=None):
+    """Convert Json file to CSV
+
+    [
+    {"User": "Alex", "Country": "US", "Age": "25"}
+    ]
+    Alex,US,25
+
+    cross tab
+    [
+    {"User": "Alex", "Country": ["US","PT"], "Age": "25"},
+    ]
+
+    User,Country,Age
+    Alex,US,25
+    Alex,PT,25
+    """
+    try:
+        fp = open(input_file, 'r')
+    except IOError as e:
+        print "I/O error({0}): {1}".format(e.errno, e.strerror)
+
+    if not is_valid_json(input_file):
+        print "invalid json file"
+        exit(1)
+
+    # set output file name and write header
+    if output_file is None:
+        output_file = str(os.path.splitext(os.path.basename(input_file))[0]) + ".csv"
+    outfile = open(output_file, 'w')
+    outfile.write(",".join(header_values))
+
+    raw_data = json.loads(fp.read())
+
+    # lines in json file
+    for item in raw_data:
+        previous_list = [""]
+        if header_values:
+            # for each line, get values corresponding to the column name values
+            for column_name in header_values:
+                new_list = []
+
+                # if column name has more than one value process ass a cross tab
+                if type(item[column_name]) is list:
+                    for child_item in item[column_name]:
+
+                        # Create new list with previous values and new cross-tab values added
+                        for old_lines in previous_list:
+                            temp = str(str(old_lines) + str(child_item) + ",")
+                            new_list.append(temp)
+                    previous_list = new_list
+
+                else:
+                    for p_strings in previous_list:
+                        new_list.append("".join(str(p_strings) + str(item[column_name]) + ","))
+                    previous_list = new_list
+
+        for lines in previous_list:
+            outfile.write("\n" + str(lines[0:-1]))
+    outfile.close()
+    fp.close()
+    os.system(" rm -r {}".format(input_file))
+    return output_file
+
+
+def xml2csv(input_file, outputfile=None, header_values=None, row_tag="row"):
+    """Convert xml to csv
+
+    <root>
+    <row>
+       <User>Alex</User>
+       <Country>US</Country>
+       <Country>PT</Country>
+       <Age>25</Age>
+    </row>
+    <row>
+       <User>Ben</User>
+       <Country>US</Country>S
+       <Age>24</Age>
+    </row>
+    </root>
+
+    User,Country,Age
+    Alex,US,25
+    Alex,Uk,25
+    Ben,US,24
+    """
+    fp = open(input_file, 'r')
+    tree = ET.parse(fp)
```

Comment:
`fp` isn't very descriptive both here and in the json function. `xml_file`, `xml_input` or anything more descriptive would be better.